src/lib/tls/asio/asio_stream.h

```diff
@@ -0,0 +1,585 @@
+            return init.result.get();
+            }
+
+         Botan::TLS::AsyncWriteOperation<typename std::decay<WriteHandler>::type, Stream>
+         op{std::move(init.completion_handler), *this, this->m_core, sent};
+
+         return init.result.get();
+         }
+
+      /**
+       * Start an asynchronous read. The function call always returns immediately.
+       *
+       * @param buffers The buffers into which the data will be read. Although the buffers object may be copied as
+       *                necessary, ownership of the underlying buffers is retained by the caller, which must guarantee
+       *                that they remain valid until the handler is called.
+       * @param handler The handler to be called when the read operation completes. The equivalent function signature of
+       *                the handler must be:
+       *                void(boost::system::error_code, std::size_t)
+       */
+      template <typename MutableBufferSequence, typename ReadHandler>
+      BOOST_ASIO_INITFN_RESULT_TYPE(ReadHandler,
+                                    void(boost::system::error_code, std::size_t))
+      async_read_some(const MutableBufferSequence& buffers, ReadHandler&& handler)
+         {
+         BOOST_ASIO_READ_HANDLER_CHECK(ReadHandler, handler) type_check;
+
+         boost::asio::async_completion<ReadHandler, void(boost::system::error_code, std::size_t)> init(handler);
+
+         AsyncReadOperation<typename std::decay<ReadHandler>::type, Stream, MutableBufferSequence>
+         op{std::move(init.completion_handler), *this, this->m_core, buffers};
+         return init.result.get();
+         }
+
+   protected:
+      size_t writePendingTlsData(boost::system::error_code& ec)
+         {
+         auto writtenBytes = boost::asio::write(m_nextLayer, this->m_core.sendBuffer(), ec);
+
+         this->m_core.consumeSendBuffer(writtenBytes);
+         return writtenBytes;
+         }
+
+      void tls_decrypt_some(boost::system::error_code& ec)
+         {
+         boost::asio::const_buffer read_buffer =
+            {
+            this->m_core.input_buffer.data(),
+            m_nextLayer.read_some(this->m_core.input_buffer, ec)
+            };
+
+         if(ec)
+            { return; }
+
+         try
+            {
+            native_handle()->received_data(static_cast<const uint8_t*>(read_buffer.data()),
+                                           read_buffer.size());
+            }
+         catch(const std::exception& ex)
+            {
+            ec = Botan::TLS::convertException();
+            }
+         }
+
+      template <typename ConstBufferSequence>
+      std::size_t tls_encrypt_some(const ConstBufferSequence& buffers,
+                                   boost::system::error_code& ec)
+         {
+         std::size_t sent = 0;
+         // NOTE: This is not asynchronous: it encrypts the data synchronously.
+         // Only writing to the socket is asynchronous.
```

Comment:
This comment is misleading when trying to understand the synchronous `write_some()` function. 

```suggestion
         // The data encrypted by native_handle()->send() is synchronously stored in the send_buffer of m_core,
         // but is not actually written to the wire, yet.
```